cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -381,10 +381,18 @@ private void createHeartbeatsFeed() throws NotificationFeedException {
       .setDescription("Stream heartbeats feed.")
       .build();
 
-    try {
-      feedManager.getFeed(streamHeartbeatsFeed);
-    } catch (NotificationFeedNotFoundException e) {
-      feedManager.createFeed(streamHeartbeatsFeed);
+    while (true) {
+      try {
+        feedManager.getFeed(streamHeartbeatsFeed);
+        return;
+      } catch (NotificationFeedNotFoundException e) {
+        feedManager.createFeed(streamHeartbeatsFeed);
+        return;
+      } catch (NotificationFeedException e) {
+        // Most probably, the dataset service is not up. We retry
+        LOG.warn("Could not subscribe to heartbeats feed", e);
+        TimeUnit.MILLISECONDS.sleep(200);
```

Comment:
since it's only called on start up, I think it is fine to sleep longer. Otherwise the log will be flooded with these.